src/mercury_engine_data_structures/formats/bmsld.py

```diff
@@ -76,73 +78,70 @@
 )
 
 CollisionObject = Struct(
-    object_type=StrId,
-    data=Switch(
+    "object_type" / StrId,
+    "data"
+    / Switch(
         construct.this.object_type,
         collision_formats,
         ErrorWithMessage(lambda ctx: f"Type {ctx.type} not known, valid types are {list(collision_formats.keys())}."),
     ),
 )
 
 BMSLD = Struct(
-    _magic=Const(b"MSLD"),
-    version=VersionAdapter("1.20.0"),
-    unk1=Int32ul,
-    unk2=Int32ul,
-    unk3=Int32ul,
-    unk4=Int32ul,
-    objects_a=make_vector(
+    "_magic" / Const(b"MSLD"),
+    "version" / VersionAdapter("1.20.0"),
+    "unk1" / Int32ul,
+    "unk2" / Int32ul,
+    "unk3" / Int32ul,
+    "unk4" / Int32ul,
+    "objects_a"
+    / make_vector(
         Struct(
-            name=StrId,
-            unk1=Hex(Int32ul),
-            unk2=Hex(Int32ul),
-            unk3=Hex(Int32ul),
-            unk4=Hex(Int32ul),
-            unk5=Hex(Int32ul),
-            unk6=Hex(Int32ul),
+            "name" / StrId,
+            "unk1" / Hex(Int32ul),
+            "unk2" / Hex(Int32ul),
+            "unk3" / Hex(Int32ul),
+            "unk4" / Hex(Int32ul),
+            "unk5" / Hex(Int32ul),
+            "unk6" / Hex(Int32ul),
         )
     ),
-    object_b=make_vector(
+    "enemy_paths"
+    / make_vector(
         Struct(
-            name=StrId,
-            unk01=Hex(Int32ul),
-            unk02=make_vector(
-                Struct(
-                    x=Float32l,
-                    y=Float32l,
-                    z=Float32l,
-                )
-            ),
+            "name" / StrId,
+            "unk01" / Hex(Int32ul),
+            "coordinates" / make_vector(CVector3D),
         )
     ),
-    objects_c=make_dict(CollisionObject),
-    objects_d=make_dict(CollisionObject),
-    objects_e=make_vector(
+    "logic_shapes" / make_dict(CollisionObject),
+    "spawn_groups" / make_dict(CollisionObject),
+    "bosses"
+    / make_vector(
         Struct(
-            name=StrId,
-            unk01=StrId,
-            unk02=Hex(Int32ul),
-            unk03=Hex(Int32ul),
-            unk04=Hex(Int32ul),
-            unk05=Hex(Int32ul),
-            unk06=Hex(Int32ul),
-            unk07=Hex(Int32ul),
-            unk08=Hex(Int32ul),
-            unk09=Float,
-            unk10=Float,
-            unk11=Hex(Int32ul),
-            unk13=StrId,
-            unk14=Hex(Int32ul),
+            "name" / StrId,
+            "unk01" / StrId,
+            "unk02" / Hex(Int32ul),
+            "unk03" / Hex(Int32ul),
+            "unk04" / Hex(Int32ul),
+            "unk05" / Hex(Int32ul),
+            "unk06" / Hex(Int32ul),
+            "unk07" / Hex(Int32ul),
+            "unk08" / Hex(Int32ul),
+            "position?" / CVector3D,
+            "type" / StrId,
+            "unk14" / Hex(Int32ul),
         )
     ),
-    actors=make_dict(ProperActor)[18],
-    sub_areas=make_vector(
+    "actors" / make_dict(ProperActor)[18],
```

Comment:
what are these 18 entries?

Reply:
I honestly don't remember what this does. @henriquegemignani do you recall?

Reply:
I did things at random until something made sense. Not a clue of any meaning

Reply:
are these like, actor layers? do all 18 have actors *in* them across the various scenarios?

Reply:
Ah right those probably are the actor layers. I don't think all of them are used though.

Reply:
can we compile a list of which layers are used, and what sorts of actors are in each layer?

Reply:
Yeah, I'm hoping we can make an int enum for layer access

Reply:
Oh I like that. Then we could reference by strinf.

Reply:
9 seems to be a general non-hostile entity layer. Because some things related to glowflies and manic miner appear on there a few times.

2 also contains spikes and other stuff (and sometimes they're also in 0, don't know what determines where exactly they go either)